pinot-query-runtime/src/main/java/org/apache/pinot/query/runtime/operator/exchange/BlockExchange.java

```diff
@@ -137,4 +165,66 @@ public void cancel(Throwable t) {
       sendingMailbox.cancel(t);
     }
   }
+
+  public SendingMailbox asSendingMailbox(String id) {
+    return new BlockExchangeSendingMailbox(id);
+  }
+
+  /**
+   * A mailbox that sends data blocks to a {@link org.apache.pinot.query.runtime.operator.exchange.BlockExchange}.
+   *
+   * BlockExchanges send data to a list of {@link SendingMailbox}es, which are responsible for sending the data
+   * to the corresponding {@link ReceivingMailbox}es. This class applies the decorator pattern to expose a BlockExchange
+   * as a SendingMailbox, open the possibility of having a BlockExchange as a destination for another BlockExchange.
+   *
+   * This is useful for example when a send operator has to send data to more than one stage. We need to broadcast the
+   * data to all the stages (the first BlockExchange). Then for each stage, we need to send the data to the
+   * corresponding workers (the inner BlockExchange). The inner BlockExchange may send data using a different
+   * distribution strategy.
+   *
+   * @see MailboxSendNode#isMultiSend()}
+   */
+  private class BlockExchangeSendingMailbox implements SendingMailbox {
```

Comment:
Neat pattern 🙂

Reply:
Yeah, it was great to find we could use decoration here. Kudos to the one that created the interface!